File: rest-service/src/main/java/mb/api/controller/GuestLinkController.java
package mb.api.controller;

import javax.ws.rs.ForbiddenException;
import javax.ws.rs.InternalServerErrorException;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.Request;

import mb.lib.db.JobDBManager;
import org.veupathdb.lib.container.jaxrs.providers.UserProvider;
import org.veupathdb.lib.container.jaxrs.server.annotations.Authenticated;
import mb.api.model.IOGuestLink;
import mb.api.controller.resources.LinkGuest;

@Authenticated
public class GuestLinkController implements LinkGuest
{
  private final Request req;

  public GuestLinkController(@Context Request req) {
    this.req = req;
  }

  @Override
  public void convGuestToUser(IOGuestLink link) {
    var user = UserProvider.lookupUser(req).orElseThrow(InternalServerErrorException::new);

    try (var db = new JobDBManager()) {
      if (db.userIsGuest(link.getGuestID())) {
        db.updateJobOwner(link.getGuestID(), user.getUserID());
      } else {
        throw new ForbiddenException();
      }
    } catch (Exception e) {
      throw new InternalServerErrorException(e);
    }
  }
}

File: rest-service/src/main/kotlin/mb/api/controller/GuestLinkController.kt
package mb.api.controller

import jakarta.ws.rs.BadRequestException
import mb.api.controller.resources.LinkGuest
import mb.api.model.IOGuestLink
import mb.lib.db.JobDBManager
import org.veupathdb.lib.container.jaxrs.providers.UserProvider
import org.veupathdb.lib.container.jaxrs.server.annotations.Authenticated
import jakarta.ws.rs.ForbiddenException
import jakarta.ws.rs.InternalServerErrorException
import jakarta.ws.rs.core.Context
import org.glassfish.jersey.server.ContainerRequest

@Authenticated
data class GuestLinkController(@Context val req: ContainerRequest): LinkGuest {

  override fun convGuestToUser(link: IOGuestLink) {
    val user = UserProvider.lookupUser(req).orElseThrow(::InternalServerErrorException)

    val guest = UserProvider.getUsersById(listOf(link.guestID)).values
      .firstOrNull()
      .let { it ?: throw BadRequestException() }
      .let { if (!it.isGuest) throw ForbiddenException() else it }

    JobDBManager().use {
      try {
        it.updateJobOwner(guest.userId, user.userId)
      } catch (e: Exception) {
        throw InternalServerErrorException(e)
      }
    }
  }
}
